src/pages/permissions/panels/EditGroupIdentitiesPanel.tsx

```diff
@@ -0,0 +1,361 @@
+import ModifiedStatusAction from "../actions/ModifiedStatusAction";
+import { pluralize } from "util/instanceBulkActions";
+import { fetchIdentities } from "api/auth-identities";
+import { LxdGroup } from "types/permissions";
+import { getCurrentIdentitiesForGroups } from "util/permissionGroups";
+import GroupIdentitiesPanelConfirmModal from "./GroupIdentitiesPanelConfirmModal";
+import PermissionIdentitiesFilter, {
+  AUTH_METHOD,
+  PermissionIdentitiesFilterType,
+  QUERY,
+} from "../PermissionIdentitiesFilter";
+import NotificationRow from "components/NotificationRow";
+
+type IdentityEditHistory = {
+  identitiesAdded: Set<string>;
+  identitiesRemoved: Set<string>;
+};
+
+interface Props {
+  selectedGroups: LxdGroup[];
+}
+
+const EditGroupIdentitiesPanel: FC<Props> = ({ selectedGroups }) => {
+  const panelParams = usePanelParams();
+  const [searchParams] = useSearchParams();
+  const notify = useNotify();
+  const [confirming, setConfirming] = useState(false);
+
+  const {
+    data: identities = [],
+    error,
+    isLoading,
+  } = useQuery({
+    queryKey: [queryKeys.identities],
+    queryFn: fetchIdentities,
+  });
+
+  const {
+    desiredState,
+    save: saveToPanelHistory,
+    undo: undoIdentitiesChange,
+  } = useEditHistory<IdentityEditHistory>({
+    initialState: {
+      identitiesAdded: new Set(),
+      identitiesRemoved: new Set(),
+    },
+  });
+
+  if (error) {
+    notify.failure("Loading panel details failed", error);
+  }
+
+  // in case if user refresh the browser while the panel is open
+  useEffect(() => {
+    if (!selectedGroups.length) {
+      panelParams.clear();
+      return;
+    }
+  }, [selectedGroups]);
+
+  const nonTlsIdentities = identities.filter(
+    (identity) => identity.authentication_method !== "tls",
+  );
+
+  const calculateIdentityAllocations = () => {
+    const { identitiesAssignedToAllGroups, identitiesAssignedToSomeGroups } =
+      getCurrentIdentitiesForGroups(selectedGroups, nonTlsIdentities);
+
+    const identitiesAllocatedToAll = new Set<string>();
+    const identitiesAllocatedToSome = new Set<string>();
+
+    for (const identity of identitiesAssignedToAllGroups) {
+      if (!desiredState.identitiesRemoved.has(identity)) {
+        identitiesAllocatedToAll.add(identity);
+      }
+    }
+
+    for (const identity of desiredState.identitiesAdded) {
+      identitiesAllocatedToAll.add(identity);
+    }
+
+    for (const identity of identitiesAssignedToSomeGroups) {
+      if (
+        identitiesAllocatedToAll.has(identity) ||
+        desiredState.identitiesRemoved.has(identity)
+      ) {
+        continue;
+      }
+
+      identitiesAllocatedToSome.add(identity);
+    }
+
+    return {
+      identitiesAllocatedToAll: Array.from(identitiesAllocatedToAll),
+      identitiesAllocatedToSome: Array.from(identitiesAllocatedToSome),
+    };
+  };
+
+  const { identitiesAllocatedToAll, identitiesAllocatedToSome } =
+    calculateIdentityAllocations();
```

Comment:
Probably best to merge this into a single calculation function that we keep in a util file?
Then we can call the function once in the global scope of the `EditGroupIdentitiesPanel` component. The interface for the util funciton could be similar to this:

```
export const calculateIdentityStatus = (
  groups: LxdGroup[],
  identities: LxdIdentity[],
  identitiesRemoved: Set<LxdIdentity>,
  identitiesAddedremoved: Set<LxdIdentity>,
): {
  identitiesInAllGroup: string[];
  identitiesInSomeGroups: string[];
  identitiesInNoGroups: string[];
  selectedIdentities: string[];
  indeterminedIdentities: string[];
} 
```

Reply:
Sure that would reduce the code in the component, although I think logic inside `calculateIdentityAllocations` is very much related to the component states whereas `getCurrentIdentitiesForGroups` is a completely independent function. When reading through the code, I find it easier to understand what's going on inside `calculateIdentityAllocations` if it's placed inside of the component since I can see the surrounding context of all the states. wdyt?

Reply:
True, we can also keep it here, as it is unlikely to be relevant to another part of the app and quite specific about this component. Maybe take over the suggested naming? I find the subtle difference of "allocated" and "assigned" could be improved on.

Reply:
Yeah I agree, maybe `getCurrentIdentitiesForGroups` can be named `getExistingIdentitiesForGroups`, then all the returned values can be `existingIdentitiesForAllGroups`, `existingIdentitiesForSomeGroups` and `existingIdentitiesForNone`?

Reply:
better! :+1: